### PR TITLE
feat: transform `hsl` to `hsla`

### LINKED ISSUE #3834: Preserve HSL colors instead of converting to RGBA (keeps colors on older browsers)
### What version of Tailwind CSS are you using?

v2.0.4

### What version of Node.js are you using?

v14.16.0

### What browser are you using?

N/A

### What operating system are you using?

macOS

### Reproduction repository

see body

I'm using HSL color definitions in a project instead of RGB/hex. With Tailwind 2.0, the `hsl()` values in my tailwind.config.js get converted to `rgba()` rules (so the opacity classes can [work their dark magic](https://adamwathan.me/composing-the-uncomposable-with-css-variables/)). However, because HSL -> RGB isn't a simple add/subtract swap, the resulting CSS uses floats for each color channel:
```js
// tailwind.config.js
module.exports = {
  theme: {
    colors: {
      offblack: 'hsl(205,95%,2%)'
} } }

```
```css
/* resulting CSS */
.text-offblack {
    --tw-text-opacity: 1;
    color: rgba(.2550000000000002,5.907499999999999,9.945,var(--tw-text-opacity));
}
```
The trouble with this is that [browser support for floats in rgba](https://caniuse.com/mdn-css_types_color_floats_in_rgb_rgba) wasn't good until the last couple years. That means people with older browsers (something like 5% of traffic) **will see NO COLORS AT ALL** on my site with Tailwind 2.0 — everything is black and white.

Can we make Tailwind keep HSL values and output `hsla()` colors instead of `rgba()`?

_Originally posted by @andronocean in https://github.com/tailwindlabs/tailwindcss/discussions/3833_

### PR DESCRIPTION
**Pull request by fedeci**
This fixes #3834 
We preserve the `hsl` colors by converting it to `hsla` instead of `rgba`.
e.g.
```diff
hsl(0, 0%, 0%)
- rgba(0, 0, 0, var(--opacity))
+ hsla(0, 0, 0, var(--opacity))
```

### CODE REVIEW

File: src/util/withAlphaVariable.js [hunk 1]
```diff
@@ -25,7 +31,9 @@ export default function withAlphaVariable({ color, property, variable }) {
   }
 
   try {
-    const [r, g, b, a] = toRgba(color)
+    const isHSL = color.startsWith('hsl')
+
+    const [i, j, k, a] = isHSL ? toHsla(color) : toRgba(color)
```
Comment (fedeci) [inline src/util/withAlphaVariable.js:36]: We use generic component names

**Comment by adamwathan** [comment]
Looking good, I noticed you still have this set as a draft so just let me know when you'd like me to look into merging 👍🏻 Thanks!

**Comment by fedeci** [comment]
Yess, I saw that `toRgba` is used in other places, so I would take a look at those too.

**Comment by fedeci** [comment]
@adamwathan it should be ready to be merged if it passes the tests

File: __tests__/plugins/gradientColorStops.test.js [hunk 2]
```diff
@@ -33,23 +34,47 @@ test('opacity variables are given to colors defined as closures', () => {
...
-					--tw-gradient-to: rgb(31,31,31)
-				}
+      .text-primary {
+        --tw-text-opacity: 1;
+        color: rgba(31, 31, 31, var(--tw-text-opacity));
+      }
+
+      .text-secondary {
+        --tw-text-opacity: 1;
+        color: hsla(10, 50, 50, var(--tw-text-opacity));
```
Comment (andronocean) [inline __tests__/plugins/gradientColorStops.test.js:44]:
```css
color: hsla(10, 50%, 50%, var(--tw-text-opacity));
```

File: __tests__/plugins/gradientColorStops.test.js [hunk 3]
```diff
@@ -33,23 +34,47 @@ test('opacity variables are given to colors defined as closures', () => {
...
+      }
+
+      .from-primary {
+        --tw-gradient-from: rgb(31, 31, 31);
+        --tw-gradient-stops: var(--tw-gradient-from), var(--tw-gradient-to, rgba(31, 31, 31, 0));
+      }
+
+      .from-secondary {
+        --tw-gradient-from: hsl(10, 50%, 50%);
+        --tw-gradient-stops: var(--tw-gradient-from), var(--tw-gradient-to, hsla(10, 50, 50, 0));
```
Comment (andronocean) [inline __tests__/plugins/gradientColorStops.test.js:58]:
```css
--tw-gradient-stops: var(--tw-gradient-from), var(--tw-gradient-to, hsla(10, 50%, 50%, 0));
```

File: __tests__/plugins/gradientColorStops.test.js [hunk 4]
```diff
@@ -33,23 +34,47 @@ test('opacity variables are given to colors defined as closures', () => {
...
+      }
+
+      .via-primary {
+        --tw-gradient-stops: var(--tw-gradient-from), rgb(31, 31, 31),
+        var(--tw-gradient-to, rgba(31, 31, 31, 0));
+      }
+
+      .via-secondary {
+        --tw-gradient-stops: var(--tw-gradient-from), hsl(10, 50%, 50%),
+        var(--tw-gradient-to, hsla(10, 50, 50, 0));
```
Comment (andronocean) [inline __tests__/plugins/gradientColorStops.test.js:68]:
```css
var(--tw-gradient-to, hsla(10, 50%, 50%, 0));
```

File: __tests__/plugins/ringWidth.test.js [hunk 5]
```diff
@@ -61,6 +61,51 @@ test('ring widths', () => {
...
+      ringColor: [],
+    },
+  }
+
+  const { utilities } = invokePlugin(plugin(), config)
+  expect(utilities).toEqual([
+    [
+      {
+        '*': {
+          '--tw-ring-color': 'hsla(10, 50, 50, 0.5)',
```
Comment (andronocean) [inline __tests__/plugins/ringWidth.test.js:88]:
```
'--tw-ring-color': 'hsla(10, 50%, 50%, 0.5)',
```

File: __tests__/withAlphaVariable.test.js [hunk 6]
```diff
@@ -118,3 +118,26 @@ test('it allows a closure to be passed', () => {
...
+  })
+  expect(
+    withAlphaVariable({
+      color: 'hsl(50, 50%, 50%)',
+      property: 'background-color',
+      variable: '--tw-bg-opacity',
+    })
+  ).toEqual({
+    '--tw-bg-opacity': '1',
+    'background-color': 'hsla(50, 50, 50, var(--tw-bg-opacity))',
```
Comment (andronocean) [inline __tests__/withAlphaVariable.test.js:141]:
```
'background-color': 'hsla(50, 50%, 50%, var(--tw-bg-opacity))',
```

**Review by andronocean — CHANGES_REQUESTED** [review]
This isn't quite right. The saturation and lightness parameters in `hsl()` or `hsla()` colors are percentages, not numbers, so they need to be followed by the `%` symbol. The tests pass, but the tests are expecting the wrong values. ([See MDN description](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#hsl_colors))

**Review by andronocean — APPROVED** [review]
Looks good to me 😎 (Thank you @fedeci !)

**Comment by andronocean** [comment]
Hey @adamwathan, any way to get this merged this week or next? It'd be really helpful 🙏

**Pull request merged**